File: Python/Algorithms/Sorting/merge_sort.py
def merge_sort(nums):
   if len(nums) <= 1:
      return nums
   
   mid_index = len(nums) // 2
   left_nums = merge_sort(nums[:mid_index])
   right_nums = merge_sort(nums[mid_index:])
   sorted = []

   left_index = 0
   right_index = 0

   while left_index < len(left_nums) and right_index < len(right_nums):
      if left_nums[left_index] < right_nums[right_index]:
         sorted.append(left_nums[left_index])
         left_index += 1
      else:
         sorted.append(right_nums[right_index])
         right_index += 1

   sorted += left_nums[left_index:]
   sorted += right_nums[right_index:]

   return sorted      
   
numbers = [2, 6, 1, 8, 10, 4, 1]
print(merge_sort(numbers))

File: GoLang/Algorithms/sorting/mergeSort.go
package sorting

import "fmt"

func MergeSort(arr []int) {
	fmt.Println(mergeSort(arr))
}

func mergeSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	mid_index := len(arr) / 2
	left_arr := mergeSort(arr[:mid_index])
	right_arr := mergeSort(arr[mid_index:])
	sorted := []int {}

	left_index := 0
	right_index := 0

	for left_index < len(left_arr) && right_index < len(right_arr) {
		if left_arr[left_index] < right_arr[right_index] {
			sorted = append(sorted, left_arr[left_index])
			left_index += 1
		} else {
			sorted = append(sorted, right_arr[right_index])
			right_index += 1
		}
	}

	sorted = append(sorted, left_arr[left_index:]...)
	sorted = append(sorted, right_arr[right_index:]...)

	return sorted
}